pfcpiface/bess.go

```diff
@@ -342,6 +343,202 @@ func (b *bess) summaryLatencyJitter(uc *upfCollector, ch chan<- prometheus.Metri
 	measureIface("Core", uc.upf.coreIface)
 }
 
+func (b *bess) readCurrentFlag(ctx context.Context) (f pb.BufferFlag, err error) {
+	any, err := anypb.New(&pb.EmptyArg{})
+	if err != nil {
+		log.Errorln("Error marshalling request", err)
+		return
+	}
+	resp, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: "flag",
+			Cmd:  "read",
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln("flag read failed!:", err)
+		return
+	}
+	flagReadResp := &pb.DoubleBufferCommandReadFlagValueResponse{}
+	if err = resp.Data.UnmarshalTo(flagReadResp); err != nil {
+		log.Errorln(err)
+		return
+	}
+	f = flagReadResp.CurrentFlag
+	return
+}
+
+func (b *bess) flipFlag(ctx context.Context) (durationNs uint64, oldFlag pb.BufferFlag, err error) {
+	// Read current flag
+	oldFlag, err = b.readCurrentFlag(ctx)
+	if err != nil {
+		return
+	}
+	// Determine new flag
+	var newFlag pb.BufferFlag
+	if oldFlag == pb.BufferFlag_FLAG_VALUE_A {
+		newFlag = pb.BufferFlag_FLAG_VALUE_B
+	} else {
+		newFlag = pb.BufferFlag_FLAG_VALUE_A
+	}
+	log.Traceln("old flag:", oldFlag, "new flag:", newFlag)
+	// Set new flag
+	req := &pb.DoubleBufferCommandSetNewFlagValueArg{NewFlag: newFlag}
+	any, err := anypb.New(req)
+	if err != nil {
+		log.Errorln("Error marshalling request", req, err)
+		return
+	}
+	resp, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: "flag",
+			Cmd:  "set",
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln("flag set failed!:", err)
+		return
+	}
+	flagSetResp := &pb.DoubleBufferCommandSetNewFlagValueResponse{}
+	if err = resp.Data.UnmarshalTo(flagSetResp); err != nil {
+		log.Errorln(err)
+		return
+	}
+	durationNs = flagSetResp.ObservationDurationNs
+	return
+}
+
+func (b *bess) readMeasurement(
+	ctx context.Context, module string, flag pb.BufferFlag, clear bool,
+) (stats pb.QosMeasureReadResponse, err error) {
+	req := &pb.QosMeasureCommandReadArg{Flag: flag, Clear: clear}
+	any, err := anypb.New(req)
+	if err != nil {
+		log.Errorln("Error marshalling request", req, err)
+		return
+	}
+	resp, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: module,
+			Cmd:  "read",
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln(module, "read failed!:", err)
+		return
+	}
+	if err = resp.Data.UnmarshalTo(&stats); err != nil {
+		log.Errorln(err)
+		return
+	}
+	return
+}
+
+func (b *bess) sessionStats(uc *upfCollector, ch chan<- prometheus.Metric) (err error) {
+	// Clearing table data with large tables is slow, let's wait for a little longer since this is
+	// non-blocking anyway.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	// Flip flag, automatically waits for pipeline to flush packets with previous mark.
+	observationDurationNs, oldFlag, err := b.flipFlag(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
Can we make this as part of read command in the (leader) module itself? So that it becomes an internal implementation detail.